core/templates/dev/head/domain/collection/CollectionValidationService.js

```diff
@@ -72,6 +72,26 @@ oppia.factory('CollectionValidationService', [
       });
     };
 
+    // Validates that the tags for the collection are in the proper format,
+    // returns true if all tags are in the correct format.
+    var validateTagFormat = function(tags) {
+      // Check to ensure that all tags follow the format specified in
+      // TAG_REGEX.
+      var tagRegex = new RegExp(GLOBALS.TAG_REGEX);
+      return tags.every(function(tag) {
+        return tag.match(tagRegex);
+      });
+    };
+
+    // Validates that the tags for the collection do not have duplicates,
+    // returns true if there are duplicates.
+    var validateDuplicateTags = function(tags) {
+      return tags.every(function(tag, idx) {
+        return tags.indexOf(tag, idx + 1) === -1 &&
+          tag === tag.trim().replace(/\s+/g, ' ');
```

Comment:
I don't think it's a good idea to put the normalization check part together with the duplicate check part, because it's not clear what the normalization is for, and normalization is conceptually a different part of the validation than duplicates. You should put the normalization into a separate validation function called `validateTagsNormalized` or something like that.